### ISSUE TITLE
Physical BCs with multiple components using AmrCore

### ISSUE THREAD

**Issue by rhouim** [AMReX-Codes/amrex#181]
I modified the Advection_AmrCore Tutorial to prepare for porting one of our codes to AMReX from the Fortran version of BoxLib.  The modified tutorial uses two components in phi_new and phi_old with a mixture of Neumann and Dirichlet BCs on the edges of the domain and the velocity field is constant in time and space.

At first, I made a vector of BCRecs as discussed in the documentation (and similar to what is done in the EB/CNS tutorial); however, the BndryFunctBase class in AMReX_PhysBCFunct.cpp does not accept a vector of BCRecs even when the multifab has multiple components. 

Currently I am using workaround where BCRec is a scalar and the boundary conditions for each component.  The logic of which boundary is which is handled on the Fortran side with modifications to phifill that change the bc array that is fed into the fillcc based on the component of phi.  This seems to work and the solution behaves as expected. Nevertheless, I hope that nothing deep in AMReX is expecting a vector of BCRecs when filling physical boundaries with multicomponent fabs, which could cause problems.

**Comment by ajnonaka** [comment]
Did you mean to say "Currently I am using a workaround where BCRec is a *vector*"?  If so, I am taking the same approach as you, which I believe is fine.  Note that all the routines in AMReX_FillPatchUtil.H (which fill newly created valid regions and ghost cells, sometimes by interpolating from coarser levels) are overloaded to take a single BCRec or a vector of BCRec.

**Comment by rhouim** [comment]
Thanks!

The workaround used a single BCRec.  The logic of how to treat the different boundaries for each component is handled in the Fortran boundary filling routines.

I also spent a little time and implemented the boundaries with a second method by modifying the AMReX_PhysBCFunct cpp and header files so that the BCFunctBase class can be constructed with a vector of BCRecs.  

Either method seems to work fine and, based on your comment, won't cause problems internally in AMReX.

**Comment by WeiqunZhang** [comment]
These class are designed for users to derive and extend.  You could write your own class derived from `PhysiBCFuncBase`.  There you can store `BndryFunctBase` and vector of `BCRec`s. `BndryFunctBase` has `virtal operator () (....., const int* bc) const`.  Is this what you are talking about not accepting vecotr of `BCRec`s?  Note that bc is `int *`, so it can be used to take the data stored in a vector.  Just pass bcrec_vector[0].data().  In your Fortran function, the argument for that is `integer, intent(in) :: bc(amrex_spacedim,2,*)`.   Here `amrex_spacedim` is from `amrex_fort_module`.  The second dimension of the bc array is 2 for lo and hi of each spatial dimension.  The last `*` is the component.  If there is only one component, you can simplify it to just `integer, intent(in) :: bc(amrex_spacedim,2)`.

**Comment by rhouim** [comment]
Thanks for the clarification Weiqun!

By not accepting a vector of BCRecs I was referring to the constructor and define routines as in AMReX_PhysBCFunc.cpp, which are written for a single BCRec.  Your solution is more-or-less exactly what I did for the second method that I discussed above.  

I have previously broken libraries in subtle ways by making similar modifications.  I just wanted to be sure that this is a correct approach.
